### PR TITLE
Use static dispatch for various layer types

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Remove various `downcast_remote_layer` and other odd `&dyn Layer` manipulations, replacing them with matching over the enum.

I think, that should be split further, but first version is created to discuss https://github.com/neondatabase/neon/pull/3260

### CODE REVIEW

File: pageserver/src/tenant/layer_map.rs [hunk 1]
```diff
@@ -20,18 +20,18 @@ use num_traits::{Bounded, Num, Signed};
...
 use utils::lsn::Lsn;
 
-use super::storage_layer::{InMemoryLayer, Layer};
+use super::storage_layer::{InMemoryLayer, Layer, PersistentLayer};
 
 ///
 /// LayerMap tracks what layers exist on a timeline.
 ///
-pub struct LayerMap<L: ?Sized> {
+pub struct LayerMap<L = PersistentLayer> {
```
Comment (hlinnaka) [inline pageserver/src/tenant/layer_map.rs:34]: I don't understand what `L = PersistentLayer` means. Does that make `L` just an alias for `PersistentLayer`? But the benchmark code uses `LayerMap<DummyMapLayer>`. How is `LayerMap<DummyMapLayer>` possible, with the constraint that `L = PeristentLayer`?

Comment (SomeoneToIgnore) [reply pageserver/src/tenant/layer_map.rs:34]: It's not a constraint, it's a default generic parameters: https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#default-generic-type-parameters-and-operator-overloading

`LayerMap<L = PersistentLayer>` means "`LayerMap` is generic, use `PersistentLayer` as a generic value if nothing else is specified explicitly"

Comment (hlinnaka) [reply pageserver/src/tenant/layer_map.rs:34]: gotcha, thanks!

**Review by hlinnaka — APPROVED** [review]
Trying to understand how this works..

With this PR, DeltaLayer, ImageLayer and RemoteLayer have functions like 'filename', local_path' etc. that mirror the functions in PersistentLayer, but they are not part of any trait. But there is a Layer trait, that contains the functions that are required by all layers. Works but feels a bit inconsistent How about having two parallel hierarchies:

```
// Trait hierarchy
trait Layer { ... };
trait PersistentLayer: Layer { ... };

// Impl hierarchy
enum LayerImpl {
    InMemory(Arc<InMemoryLayer>),
    Persistent(PersistentLayerImpl),
}
impl Layer for LayerImpl {
    ...
}
enum PersistentLayerImpl {
    Delta(Arc<DeltaLayer>),
    Image(Arc<ImageLayer>),
    Remote(Arc<RemoteLayer>),
}
impl PersistentLayer for PersistentLayerImpl {
    ...
}
```
With this, you have the concrete types for both Layer and PersistentLayer, for when you need to store those things. And you also have the traits, so that you can have generics over all PersistentLayer, for example.

Or maybe we don't need the traits at all anymore. We could have just the Impl hierarchy. Although I feel the traits are nice as documentation for which functions are common between all Layers and which are are common between all PersistentLayers, if nothing else.

**Comment by SomeoneToIgnore** [comment]
I have another draft with more advanced generics: https://github.com/neondatabase/neon/pull/3263
It continues the path proposed here and, while a bit rough still (hence the Draft status, it has some impl duplication that I don't like), it finally brings the typization on the level I like: it knows about Delta, Image and Historic layers and their properties (can be remote).

I will work on polishing that, so I think, the trait hierarchy is not needed, unless I'll find some good way to share the code into via such trait impl for enums.

File: pageserver/src/tenant/storage_layer.rs [hunk 2]
```diff
@@ -83,6 +84,11 @@ pub enum ValueReconstructResult {
 /// Supertrait of the [`Layer`] trait that captures the bare minimum interface
 /// required by [`LayerMap`].
 pub trait Layer: Send + Sync {
```
Comment (bojanserafimov) [inline pageserver/src/tenant/storage_layer.rs:86]: A bit unrelated to the PR (feel free to ignore), but some thoughts on the direction of this code:

Judging by the trait docstring, this trait is to `LayerMap` what `Hash` is to `HashMap`. The trait is about key types that are good enough to be indexed by the layer map.

But the name `Layer` implies something more, and talks about the value contained too. The method `get_value_reconstruct_data` is definitely not in the `bare minimum interface required by LayerMap`, but it is a common method that all layers should implement, so it's here.

A possible fix is to rename this trait to `LayerRange` (or `LayerMetadata`, or `LayerDescriptor`), and add a separate trait for `LayerContent`, containing `get_value_reconstruct_data` and `dump`. After this change we won't have anything named `Layer`, which is IMO a good thing. If there was such a thing it should be an enum, not a trait. Or if it is a trait it should just be `pub trait Layer: LayerRange + LayerContent {}`.

Then the only outliers remaining are `get_tenant_id` and `get_timeline_id`. I don't know why these are necessary. The owner of the layer is a timeline, so whoever is using the layer already has a way to get the tenant and timeline id. Maybe they're here just for convenience? Do they have to be part of the trait?

**Review by problame — COMMENTED** [review]
This seems better than the hacky `InMemoryOrHistoricLayer` in #3260 , but, [as I pointed out in that PR's comments](https://github.com/neondatabase/neon/pull/3260#issuecomment-1370752407), I don't get why we need this _right now_.

If I'm right and we don't need it for `RequestContext`, let's park this until

1. we have shipped on-demand download to prod and
2. we have merged the `RequestContext` work #3228

**Comment by SomeoneToIgnore** [comment]
Yes, that can wait.
Bojan's comments are really useful and I need to work more on the proposals.
Also, https://github.com/neondatabase/neon/pull/3263 gets even better when I worked on it more, so I think I'll try to get that in too, eventually.

**Comment by SomeoneToIgnore** [comment]
I have implemented my ideas and PR's suggestions in this PR: https://github.com/neondatabase/neon/pull/3263

Way more satisfied with the fact, that we're now able to distinguish between remote, delta and image layers better on the type level.
Splitting `Trait` into subtraits and renaming them helps with that too, thank you for suggestions.
I will add more comments (and code simplifications, if find any) and mark that PR ready for the review.

**Pull request closed without merging**